### PR TITLE
Choose which specific people instead of a role in question recipients

### LINKED ISSUE #31: Select individual recipients in Questions workflow
## Is your feature request related to a problem? Please describe.
Administrators ask for the selection of individual recipients instead of a role (_committee_ or _service_) when updating the Question state (_evaluating_).

_see **Additional context** for more details on the module workflow_

## Describe the solution you'd like
When moving the Question in the workflow (updating state / recipient), user can select multiple users as recipient (instead of all the users with the targeted role) : 
- [x] Add a **new attribute `recipient_ids`** on the `Question` model to store multiple user ID.
- [x] Update the recipient sub-form on the **Question admin edit form** with a `select` control of the targeted role users (_see screenshots in **Additional context**_)
- [x] Restrict the **Question admin index** list with the selected recipient (_see details in **Additional context**_)
- [x] Update Notifications recipients according to the selected recipients.

➡️ ⚠️ **These enhancements should be made from the [`feature/upstream_moderation`](https://github.com/OpenSourcePolitics/decidim-questions/tree/feature/upstream_moderation) branch**

> Note : `Permissions` will be handle in another issue.

## Describe alternatives you've considered
"Component roles" and a more advanced workflow management system were discussed but the complexity and impact on the whole platform did not fit the planning related to this enhancement. 

On the Question workflow, the `committee` and `service` roles are used to distribute Questions to corresponding user groups when the Question is being evaluated / answered. This impact Question visibility, actions and notifications for this users. 

Add a **new attribute `recipient_ids`** on the `Question` model to store multiple user IDs. 

### Question admin index
On the component admin index, `service` users are **only** able to see the Questions whom they are **recipient** for. 
> No change for Admins and `committee` users : they are able to see everything.

### Question admin edit form
Prior to this enhancement, the Question admin edit form was updated to be able to : 
- Select the state of the Question
- When the state was "evaluating", select the recipient of the Questions (`none`/`committee`/`service`)
- When the state was "accepted" or "rejected", directly answer the Question from this form

This enhancement add a new step :  
When the recipient is a role (`committee`/`service`), display a component to select users (filtered by the targeted role). The `select` component must be set with options : 
- multiple selection enabled
- first and default option to `<all>` (which will select all users in the list)

Capture d’écran 2019-05-09 à 15 36 17

Capture d’écran 2019-05-09 à 15 40 48

## Update Notifications recipients
Notifications recipients defined in the `Decidim::Questions::Admin::UpdateQuestion` (`app/commands/decidim/questions/admin/update_question.rb`) command must be updated to restrict them only to selected users.

### PR DESCRIPTION
**Pull request by oriolgual**
#### :tophat: What? Why?

Adds an option to choose the recipients for a question.

#### :pushpin: Related Issues
- Fixes #31

#### :clipboard: Subtasks
- [x] Choose committee users
- [x] Choose service users
- [x] Notify selected users

### :camera: Screenshots
Demo

### CODE REVIEW

File: app/assets/javascripts/decidim/questions/admin/questions_form.js.es6 [hunk 1]
```diff
@@ -26,6 +41,19 @@ $(() => {
           answer.removeClass('hide');
         }
       })
+
+      recipients.change((e) => {
```
Comment (oriolgual) [inline app/assets/javascripts/decidim/questions/admin/questions_form.js.es6]: DRY this

File: app/commands/decidim/questions/admin/update_question.rb [hunk 2]
```diff
@@ -97,22 +99,14 @@ def versioned_attributes_changed?
         end
 
         def notify_recipients
-          return if (question.previous_changes.keys & %w[recipient]).empty?
+          return if (question.previous_changes.keys & %w[recipient recipient_ids]).empty?
```
Comment (oriolgual) [inline app/commands/decidim/questions/admin/update_question.rb:106]: Add specs

File: app/forms/decidim/questions/admin/question_form.rb [hunk 3]
```diff
@@ -23,6 +23,8 @@ class QuestionForm < Decidim::Form
         attribute :suggested_hashtags, Array[String]
         attribute :state, String
         attribute :recipient, String
+        attribute :committee_users_ids, Array[Integer]
+        attribute :service_users_ids, Array[Integer]
```
Comment (oriolgual) [inline app/forms/decidim/questions/admin/question_form.rb]: Add validations

File: app/forms/decidim/questions/admin/question_form.rb [hunk 4]
```diff
@@ -47,6 +49,12 @@ def map_model(model)
           self.category_id = model.categorization.decidim_category_id
           self.scope_id = model.decidim_scope_id
 
+          if model.recipient == "committee"
+            self.committee_users_ids = model.recipient_ids
+          elsif model.recipient == "service"
+            self.service_users_ids = model.recipient_ids
```
Comment (oriolgual) [inline app/forms/decidim/questions/admin/question_form.rb]: Add specs

File: app/views/decidim/questions/admin/questions/_form.html.erb [hunk 5]
```diff
@@ -70,6 +70,22 @@
         </div>
       </div>
     </div>
+    <% if @form.service_users&.any? %>
+      <div class="row column hide" id="question_service_users_wrapper">
+        <%= form.select :service_users_ids,
+          @form.service_users.map{|user| [user.name, user.id]}.prepend(["All", "all"]),
```
Comment (oriolgual) [inline app/views/decidim/questions/admin/questions/_form.html.erb]: i18n

File: app/views/decidim/questions/admin/questions/_form.html.erb [hunk 6]
```diff
@@ -70,6 +70,22 @@
...
+        { include_blank: false },
+        { multiple: true, class: "chosen-select" } %>
+      </div>
+    <% end %>
+    <% if @form.committee_users&.any? %>
+      <div class="row column hide" id="question_committee_users_wrapper">
+        <%= form.select :committee_users_ids,
+            @form.committee_users.map{|user| [user.name, user.id]}.prepend(["All", "all"]),
```
Comment (oriolgual) [inline app/views/decidim/questions/admin/questions/_form.html.erb]: i18n

File: app/views/decidim/questions/admin/questions/_form.html.erb [hunk 7]
```diff
@@ -70,6 +70,22 @@
...
+    <% if @form.service_users&.any? %>
+      <div class="row column hide" id="question_service_users_wrapper">
+        <%= form.select :service_users_ids,
+          @form.service_users.map{|user| [user.name, user.id]}.prepend(["All", "all"]),
+        { include_blank: false },
+        { multiple: true, class: "chosen-select" } %>
+      </div>
+    <% end %>
+    <% if @form.committee_users&.any? %>
+      <div class="row column hide" id="question_committee_users_wrapper">
```
Comment (oriolgual) [inline app/views/decidim/questions/admin/questions/_form.html.erb]: Extract to helper?

File: db/migrate/20190514140811_add_recipient_ids_to_questions.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class AddRecipientIdsToQuestions < ActiveRecord::Migration[5.2]
+  def change
+    add_column :decidim_questions_questions, :recipient_ids, :integer, array: true, default: [], null: false
```
Comment (oriolgual) [inline db/migrate/20190514140811_add_recipient_ids_to_questions.rb:5]: Add migration for existing questions

**Comment by oriolgual** [comment]
Ready to review @josepjaume & @moustachu

File: lib/decidim/questions/component.rb [hunk 9]
```diff
@@ -163,6 +163,33 @@
       global = nil
     end
 
+    email = "committee-user-#{participatory_space.underscored_name}-#{participatory_space.id}@example.org"
+    name = "#{Faker::Name.name} #{participatory_space.id} #{n} committee"
```
Comment (armandfardeau) [inline lib/decidim/questions/component.rb]: What does the "n" stand for?
`name = "#{Faker::Name.name} #{participatory_space.id} #{n} committee"`

File: lib/decidim/questions/component.rb [hunk 10]
```diff
@@ -163,6 +163,33 @@
...
+
+    committee_user = Decidim::User.find_or_initialize_by(email: email)
+    author.update!(
```
Comment (armandfardeau) [inline lib/decidim/questions/component.rb]: Shouldn't it be "committee_user"?
```suggestion
    committee_user.update!(
```

File: lib/decidim/questions/component.rb [hunk 11]
```diff
@@ -163,6 +163,33 @@
...
+      name: name,
+      nickname: Faker::Twitter.unique.screen_name,
+      organization: component.organization,
+      tos_agreement: "1",
+      confirmed_at: Time.current
+    )
+    Decidim::ParticipatoryProcessUserRole.create!(role: "committee", participatory_process: participatory_space, user: committee_user)
+
+    service_user = Decidim::User.find_or_initialize_by(email: email)
+    author.update!(
```
Comment (armandfardeau) [inline lib/decidim/questions/component.rb]: Shouldn't it be "service_user"?

```suggestion
    service_user.update!(
```

**Comment by armandfardeau** [comment]
@oriolgual Many thanks for your contribution, @moustachu This is okay for me. I updated translations in French, but cannot manage to run the seed? Could you check?

File: lib/decidim/questions/component.rb (hunk 9 above)
Comment (oriolgual) [reply lib/decidim/questions/component.rb]: It was a mistake, sorry!

**Comment by oriolgual** [comment]
Seeds should be working now!

**Comment by armandfardeau** [comment]
@oriolgual Thanks. It will do the job !

**Pull request merged**